### PR TITLE
Make secret complex in order to make the captcha stronger

### PR DESCRIPTION
**Pull request by vvvllll**
This module is vulnerable by default if you don't put some more data inside "secret" (like the client ip, a timestamp etc).

This is an attempt to make the "secret" a complex value, however it's corrupting the code var (it contains arbitrary data appended after this patch).
If you can check to implement it and fix my code.

### CODE REVIEW

File: ngx_http_captcha_module.c [hunk 1]
```diff
@@ -54,7 +54,10 @@ static ngx_int_t ngx_http_captcha_handler(ngx_http_request_t *r) {
     }
     ngx_md5_t md5;
     (void)ngx_md5_init(&md5);
-    (void)ngx_md5_update(&md5, (const void *)location->secret.data, location->secret.len);
+    ngx_str_t               secretval;
+    ngx_http_complex_value(r, location->secret, &secretval);
```
Comment (RekGRpth) [inline ngx_http_captcha_module.c:58]:
```suggestion
    if (ngx_http_complex_value(r, location->secret, &secret) != NGX_OK) { ngx_log_error(NGX_LOG_ERR, r->connection->log, 0, "ngx_http_complex_value != NGX_OK"); return NGX_HTTP_INTERNAL_SERVER_ERROR; }
```

File: ngx_http_captcha_module.c [hunk 2]
```diff
@@ -54,7 +54,10 @@ static ngx_int_t ngx_http_captcha_handler(ngx_http_request_t *r) {
     }
     ngx_md5_t md5;
     (void)ngx_md5_init(&md5);
-    (void)ngx_md5_update(&md5, (const void *)location->secret.data, location->secret.len);
+    ngx_str_t               secretval;
```
Comment (RekGRpth) [inline ngx_http_captcha_module.c:57]:
```suggestion
    ngx_str_t secret;
```

File: ngx_http_captcha_module.c [hunk 3]
```diff
@@ -54,7 +54,10 @@ static ngx_int_t ngx_http_captcha_handler(ngx_http_request_t *r) {
...
+    ngx_http_complex_value(r, location->secret, &secretval);
+    ngx_log_error(NGX_LOG_WARN, r->connection->log, 0, "secret0 %s", code);
```
Comment (RekGRpth) [inline ngx_http_captcha_module.c:59]:
```suggestion
```

File: ngx_http_captcha_module.c [hunk 4]
```diff
@@ -54,7 +54,10 @@ static ngx_int_t ngx_http_captcha_handler(ngx_http_request_t *r) {
...
+    (void)ngx_md5_update(&md5, (const void *)&secretval, secretval.len);
```
Comment (RekGRpth) [inline ngx_http_captcha_module.c:60]:
```suggestion
    (void)ngx_md5_update(&md5, (const void *)&secret.data, secret.len);
```

File: ngx_http_captcha_module.c [hunk 5]
```diff
@@ -235,10 +238,11 @@ static char *ngx_http_captcha_merge_loc_conf(ngx_conf_t *cf, void *parent, void
     ngx_conf_merge_str_value(conf->csrf, prev->csrf, "csrf");
     ngx_conf_merge_str_value(conf->font, prev->font, "/usr/local/share/fonts/NimbusSans-Regular.ttf");
     ngx_conf_merge_str_value(conf->name, prev->name, "Captcha");
-    ngx_conf_merge_str_value(conf->secret, prev->secret, "secret");
+    if (conf->secret == NULL) {
```
Comment (RekGRpth) [inline ngx_http_captcha_module.c:241]:
```suggestion
    if (!conf->secret) conf->secret = prev->secret;
```

File: ngx_http_captcha_module.c [hunk 6]
```diff
@@ -235,10 +238,11 @@ static char *ngx_http_captcha_merge_loc_conf(ngx_conf_t *cf, void *parent, void
...
+        conf->secret = prev->secret;
```
Comment (RekGRpth) [inline ngx_http_captcha_module.c:242]:
```suggestion
```

File: ngx_http_captcha_module.c [hunk 7]
```diff
@@ -235,10 +238,11 @@ static char *ngx_http_captcha_merge_loc_conf(ngx_conf_t *cf, void *parent, void
...
+    }
```
Comment (RekGRpth) [inline ngx_http_captcha_module.c:243]:
```suggestion
```

File: ngx_http_captcha_module.c [hunk 8]
```diff
@@ -235,10 +238,11 @@ static char *ngx_http_captcha_merge_loc_conf(ngx_conf_t *cf, void *parent, void
...
     if (conf->size > conf->height) return "captcha size is too large";
     if (!conf->name.len) return "captcha name cannot be empty";
-    if (!conf->secret.len) return "captcha secret cannot be empty";
```
Comment (RekGRpth) [inline ngx_http_captcha_module.c:241]: this line should be not deleted!

**Review by RekGRpth — CHANGES_REQUESTED** [review]

File: ngx_http_captcha_module.c (hunk 8 above)
Comment (vvvllll) [reply ngx_http_captcha_module.c:241]: It doesn't compile with it

ngx_http_captcha_module//ngx_http_captcha_module.c: In function 'ngx_http_captcha_merge_loc_conf':
ngx_http_captcha_module//ngx_http_captcha_module.c:244:22: error: 'conf->secret' is a pointer; did you mean to use '->'?
  244 |     if (!conf->secret.len) return "captcha secret cannot be empty";
      |                      ^
      |                      ->

**Review by vvvllll — COMMENTED** [review]
Still broken for me, "code" contains invalid value after adding the line 

    if (ngx_http_complex_value(r, location->secret, &secret) != NGX_OK) { ngx_log_error(NGX_LOG_ERR, r->connection->log, 0, "ngx_http_complex_value != NGX_OK"); return NGX_HTTP_INTERNAL_SERVER_ERROR; }

it contains arbitrary data, eg if the captcha is "abcd", and then i print "code" in nginx logs after this line, it will contain part of the nginx configuration file so there might have a memory error somewhere

This also makes the md5 always wrong

File: ngx_http_captcha_module.c (hunk 3 above)
Comment (vvvllll) [reply ngx_http_captcha_module.c:59]: Still corrupts "code"

**Comment by RekGRpth** [comment]
I apply you commit and add fix

**Pull request closed without merging**